components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py

```diff
@@ -0,0 +1,86 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+from typing import List, Optional
+
+import torch
+from tensorrt_llm.sampling_params import LogitsProcessor
+
+from dynamo.logits_processing import BaseLogitsProcessor
+
+logger = logging.getLogger(__name__)
+
+
+class TrtllmDynamoLogitsAdapter(LogitsProcessor):
+    """
+    Adapter that wraps Dynamo BaseLogitsProcessor instances to work with TensorRT-LLM's logits processor interface.
+
+    Inherits from tensorrt_llm.LogitsProcessor and implements the required interface:
+    __call__(self, req_ids: int, logits: torch.Tensor, ids: List[List[int]], stream_ptr, client_id: Optional[int])
+
+    This adapter maintains per-request state and converts between the interfaces.
+    """
+
+    def __init__(self, processor: BaseLogitsProcessor):
+        super().__init__()
+        self.processor = processor
+
+    def __call__(
+        self,
+        req_ids: int,
+        logits: torch.Tensor,
+        ids: List[List[int]],
+        stream_ptr,
+        client_id: Optional[int] = None,
+    ):
+        """
+        TensorRT-LLM logits processor interface.
+
+        Args:
+            req_ids: Request identifier
+            logits: Logits tensor for current step
+            ids: List of token sequences (batch of sequences)
+            stream_ptr: CUDA stream pointer
+            client_id: Optional client identifier
+
+        Returns:
+            Modified logits tensor (in-place modification expected by TRT-LLM)
+        """
+        print(f"Shapes: logits {logits.shape}, ids {ids}")
+        try:
+            for ids_req, logits_req in zip(ids, logits):
+                if logits_req.shape[0] != 1:
+                    raise ValueError(
+                        "Logits processing with beam width > 1 is not supported"
+                    )
+                # Remove dimension 0 from logits_req
+                modified_logits = self.processor(ids_req, logits_req.reshape(-1))
+
+                # TRT-LLM expects in-place modification
+                logits.copy_(modified_logits)
```

Comment:
Have you tried putting those two ops in a @torch.compile function to see if we can have a JIT kernel fusing these two? Or some other ways to limit the overhead / number of kernels

Reply:
If torch compile is not successful it would fall back to the regular eager execution so it should be fine. I think we should focus on performance by default especially in these core components. Would be bad UX to ask users to modify or extend the code we provide, if they want performance.

Reply:
Right but since the extra op was in the adapter, then the user-provided callback would not have the opportunity to fuse its ops with the ones the adapter was doing. Now that you removed the copy it should be good but let's keep it in mind if we later add some more logic on the adapter side